File: shared/api-types.ts
/**
 * Shared API contract between Argus frontend (Next.js + Tambo) and backend (FastAPI + mcp).
 *
 * The frontend's API client imports these types. The backend emits card descriptors that
 * conform to the Zod schemas in research/14-zod-schemas.md.
 *
 * This file is the source of truth. If you change a type here, both sides must be updated.
 */

import { z } from 'zod';

// =============================================================================
// 7 Card Schemas (the contract between card_renderer.py and the React components)
// Per research/14-zod-schemas.md
// =============================================================================

export const StatCardProps = z.object({
  label: z.string().describe('Human-readable label, e.g. "Active users"'),
  value: z.number().describe('The numeric value'),
  unit: z.enum(['', 'count', 'percent', 'usd', 'eur', 'gbp', 'days', 'hours']).optional(),
  delta: z.number().optional().describe('Optional change from previous period (signed)'),
  deltaWindow: z.enum(['day', 'week', 'month']).optional(),
  trend: z.enum(['up', 'down', 'flat']).optional(),
  comparisonText: z.string().optional(),
});

export const TimeSeriesCardProps = z.object({
  title: z.string(),
  series: z.array(z.object({
    name: z.string(),
    points: z.array(z.object({
      t: z.string().describe('ISO 8601 timestamp'),
      v: z.number(),
    })),
    color: z.enum(['primary', 'secondary', 'success', 'warning', 'danger']).optional(),
  })).min(1),
  xAxisLabel: z.string().optional(),
  yAxisLabel: z.string().optional(),
  granularity: z.enum(['hour', 'day', 'week', 'month']).optional(),
  showAnomalies: z.boolean().optional(),
});

export const BarChartCardProps = z.object({
  title: z.string(),
  orientation: z.enum(['vertical', 'horizontal']).default('vertical'),
  bars: z.array(z.object({
    label: z.string(),
    value: z.number(),
    color: z.enum(['primary', 'secondary', 'success', 'warning', 'danger']).optional(),
  })).min(1).max(20),
  xAxisLabel: z.string().optional(),
  yAxisLabel: z.string().optional(),
  sortBy: z.enum(['value-desc', 'value-asc', 'label-asc']).default('value-desc'),
  showValues: z.boolean().default(true),
});

export const HeatmapCardProps = z.object({
  title: z.string(),
  rowLabels: z.array(z.string()).min(2).max(20),
  colLabels: z.array(z.string()).min(2).max(20),
  values: z.array(z.array(z.number())),
  colorScale: z.enum(['sequential', 'diverging']).default('sequential'),
  colorDomain: z.tuple([z.number(), z.number()]).optional(),
  cellFormat: z.enum(['count', 'percent', 'usd']).default('count'),
  showRowLabels: z.boolean().default(true),
  showColLabels: z.boolean().default(true),
});

export const TableCardProps = z.object({
  title: z.string(),
  columns: z.array(z.object({
    key: z.string(),
    label: z.string(),
    format: z.enum(['text', 'number', 'percent', 'usd', 'date', 'datetime']).default('text'),
    sortable: z.boolean().default(true),
    align: z.enum(['left', 'center', 'right']).default('left'),
  })).min(1).max(10),
  rows: z.array(z.record(z.union([z.string(), z.number(), z.null()]))).min(1).max(1000),
  pageSize: z.number().min(10).max(100).default(25),
  enableSearch: z.boolean().default(true),
  enableExport: z.boolean().default(true),
});

export const SummaryCardProps = z.object({
  title: z.string(),
  summary: z.string(),
  findings: z.array(z.object({
    text: z.string(),
    severity: z.enum(['info', 'warning', 'critical']).default('info'),
    metric: z.string().optional(),
    delta: z.number().optional(),
  })).max(5),
  suggestedActions: z.array(z.string()).max(3).optional(),
  relatedCard: z.union([
    StatCardProps,
    TimeSeriesCardProps,
    BarChartCardProps,
  ]).optional(),
});

export const ErrorCardProps = z.object({
  title: z.string().default('Something went wrong'),
  message: z.string(),
  errorCode: z.string().optional(),
  technicalDetails: z.string().optional(),
  isRetryable: z.boolean().default(true),
  isReadOnlyViolation: z.boolean().default(false),
  isFullPage: z.boolean().default(false),
  guidance: z.string().optional(),
});

export const CardName = z.enum([
  'StatCard',
  'TimeSeriesCard',
  'BarChartCard',
  'HeatmapCard',
  'TableCard',
  'SummaryCard',
  'ErrorCard',
]);

export const CardDescriptor = z.object({
  componentName: CardName,
  props: z.record(z.unknown()),
});

export type StatCardPropsType = z.infer<typeof StatCardProps>;
export type TimeSeriesCardPropsType = z.infer<typeof TimeSeriesCardProps>;
export type BarChartCardPropsType = z.infer<typeof BarChartCardProps>;
export type HeatmapCardPropsType = z.infer<typeof HeatmapCardProps>;
export type TableCardPropsType = z.infer<typeof TableCardProps>;
export type SummaryCardPropsType = z.infer<typeof SummaryCardProps>;
export type ErrorCardPropsType = z.infer<typeof ErrorCardProps>;
export type CardNameType = z.infer<typeof CardName>;
export type CardDescriptorType = z.infer<typeof CardDescriptor>;

// =============================================================================
// API Request/Response types
// Per research/15-routes.md
// =============================================================================

// POST /api/v1/connect
export const ConnectRequest = z.object({
  connection_string: z.string().min(1),
  acknowledged_risks: z.array(z.string()).optional(),
});
export const ConnectResponse = z.object({
  session_token: z.string(),
  status: z.enum(['probing', 'ready', 'error']),
  error: ErrorCardProps.optional(),
});
export type ConnectRequestType = z.infer<typeof ConnectRequest>;
export type ConnectResponseType = z.infer<typeof ConnectResponse>;

// GET /api/v1/probe/{token}
export const ProbeResponse = z.object({
  collections: z.number(),
  database_name: z.string(),
  mongo_version: z.string().optional(),
  cluster_tier: z.string().optional(),
  sample_status: z.enum(['pending', 'sampling', 'ready', 'error']),
});
export type ProbeResponseType = z.infer<typeof ProbeResponse>;

// GET /api/v1/sample/{token}
export const SampleResponse = z.object({
  collections: z.array(z.object({
    name: z.string(),
    doc_count: z.number(),
    sample_fields: z.array(z.string()),
  })),
});
export type SampleResponseType = z.infer<typeof SampleResponse>;

// POST /api/v1/plan
export const PlanRequest = z.object({
  session_token: z.string(),
  collections: z.array(z.string()).min(1).max(20),
  modules: z.array(z.enum(['funnel', 'cohort', 'rfm', 'attribution', 'anomaly'])).min(1).max(5),
});
export const PlanResponse = z.object({
  plan_id: z.string(),
  plan: z.array(z.object({
    module: z.string(),
    collection: z.string(),
    mql_pipeline: z.array(z.record(z.unknown())),
    estimated_runtime_s: z.number().optional(),
  })),
});
export type PlanRequestType = z.infer<typeof PlanRequest>;
export type PlanResponseType = z.infer<typeof PlanResponse>;

// POST /api/v1/render (SSE stream)
export const RenderRequest = z.object({
  session_token: z.string(),
  plan_id: z.string(),
});
export const RenderEvent = z.union([
  z.object({ event: z.literal('card'), data: CardDescriptor }),
  z.object({ event: z.literal('progress'), data: z.object({ completed: z.number(), total: z.number() }) }),
  z.object({ event: z.literal('error'), data: ErrorCardProps }),
  z.object({ event: z.literal('done'), data: z.object({ plan_id: z.string() }) }),
]);
export type RenderRequestType = z.infer<typeof RenderRequest>;
export type RenderEventType = z.infer<typeof RenderEvent>;

// GET /api/v1/dashboard
export const DashboardResponse = z.object({
  layout: z.object({
    lg: z.array(z.object({ i: z.string(), x: z.number(), y: z.number(), w: z.number(), h: z.number() })),
    md: z.array(z.object({ i: z.string(), x: z.number(), y: z.number(), w: z.number(), h: z.number() })),
    sm: z.array(z.object({ i: z.string(), x: z.number(), y: z.number(), w: z.number(), h: z.number() })),
    xs: z.array(z.object({ i: z.string(), x: z.number(), y: z.number(), w: z.number(), h: z.number() })),
  }),
  cards: z.array(CardDescriptor),
});
export type DashboardResponseType = z.infer<typeof DashboardResponse>;

// PUT /api/v1/dashboard/layout
export const LayoutRequest = z.object({
  session_token: z.string(),
  layout: DashboardResponse.shape.layout,
});
export const LayoutResponse = z.object({ ok: z.boolean() });
export type LayoutRequestType = z.infer<typeof LayoutRequest>;
export type LayoutResponseType = z.infer<typeof LayoutResponse>;

// POST /api/v1/cards
export const AddCardRequest = z.object({
  session_token: z.string(),
  module: z.enum(['funnel', 'cohort', 'rfm', 'attribution', 'anomaly']),
  collection: z.string().optional(),
  params: z.record(z.unknown()).optional(),
});
export const AddCardResponse = z.object({
  card_id: z.string(),
  card: CardDescriptor,
});
export type AddCardRequestType = z.infer<typeof AddCardRequest>;
export type AddCardResponseType = z.infer<typeof AddCardResponse>;

// POST /api/v1/refresh
export const RefreshRequest = z.object({
  session_token: z.string(),
});
export const RefreshResponse = z.object({
  refreshed_count: z.number(),
  error_count: z.number(),
});
export type RefreshRequestType = z.infer<typeof RefreshRequest>;
export type RefreshResponseType = z.infer<typeof RefreshResponse>;

// GET /api/v1/health
export const HealthResponse = z.object({
  status: z.enum(['ok', 'degraded', 'down']),
  uptime_s: z.number(),
  version: z.string(),
  mcp_subprocesses: z.object({
    active: z.number(),
    max: z.number(),
  }),
});
export type HealthResponseType = z.infer<typeof HealthResponse>;

// =============================================================================
// Error envelope
// Per research/19-error-states.md
// =============================================================================

export const ErrorCode = z.enum([
  'CONNECTION_FAILED',
  'INVALID_INPUT',
  'RATE_LIMITED',
  'READ_ONLY_VIOLATION',
  'MQL_EXECUTION_FAILED',
  'LLM_HALLUCINATION',
  'SCHEMA_SAMPLE_FAILED',
  'MCP_SUBPROCESS_FAILED',
  'UNKNOWN',
]);
export type ErrorCodeType = z.infer<typeof ErrorCode>;

export const ErrorResponse = z.object({
  error: z.object({
    code: ErrorCode,
    message: z.string(),
    technicalDetails: z.string().optional(),
    isRetryable: z.boolean(),
    guidance: z.string().optional(),
  }),
});
export type ErrorResponseType = z.infer<typeof ErrorResponse>;

File: backend/argus/models/api_types.py
"""Pydantic v2 models for the 11 API endpoints.

These mirror the request/response shapes in `shared/api-types.ts` (the
source of truth). Field names match the Zod field names exactly —
**most are snake_case, but a few (technicalDetails, isRetryable,
mcpSubprocesses) are camelCase** per the Zod schema. The user said:
"Pydantic field names MUST match the Zod field names."

For camelCase fields we use ``Field(alias=...)`` + ``populate_by_name=True``
so both the alias and the snake_case Python name are accepted on
input. Output uses the alias (the wire format) by default.
"""

from __future__ import annotations

from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from argus.models.card import CardDescriptor, ErrorCardProps

# ---------------------------------------------------------------------------
# Error envelope
# ---------------------------------------------------------------------------


class ErrorCode(str, Enum):
    """Mirrors the `ErrorCode` enum in shared/api-types.ts."""

    CONNECTION_FAILED = "CONNECTION_FAILED"
    INVALID_INPUT = "INVALID_INPUT"
    RATE_LIMITED = "RATE_LIMITED"
    READ_ONLY_VIOLATION = "READ_ONLY_VIOLATION"
    MQL_EXECUTION_FAILED = "MQL_EXECUTION_FAILED"
    LLM_HALLUCINATION = "LLM_HALLUCINATION"
    SCHEMA_SAMPLE_FAILED = "SCHEMA_SAMPLE_FAILED"
    MCP_SUBPROCESS_FAILED = "MCP_SUBPROCESS_FAILED"
    UNKNOWN = "UNKNOWN"


class ErrorBody(BaseModel):
    """The ``error`` object inside an ``ErrorResponse`` envelope.

    Per the Zod schema, ``technicalDetails`` and ``isRetryable`` are
    camelCase on the wire; the rest are snake_case.
    """

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    code: ErrorCode
    message: str
    technical_details: str | None = Field(default=None, alias="technicalDetails")
    is_retryable: bool = Field(..., alias="isRetryable")
    guidance: str | None = None


class ErrorResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    error: ErrorBody


# ---------------------------------------------------------------------------
# POST /api/v1/connect
# ---------------------------------------------------------------------------


class ConnectRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    connection_string: str = Field(..., min_length=1)
    acknowledged_risks: list[str] | None = Field(default=None, alias="acknowledgedRisks")


class ConnectStatus(str, Enum):
    PROBING = "probing"
    READY = "ready"
    ERROR = "error"


class ConnectResponse(BaseModel):
    """snake_case on the wire: ``session_token``."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str
    status: ConnectStatus
    error: ErrorCardProps | None = None


# ---------------------------------------------------------------------------
# GET /api/v1/probe/{token}
# ---------------------------------------------------------------------------


class SampleStatus(str, Enum):
    PENDING = "pending"
    SAMPLING = "sampling"
    READY = "ready"
    ERROR = "error"


class ProbeResponse(BaseModel):
    """snake_case on the wire."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    collections: int
    database_name: str
    mongo_version: str | None = None
    cluster_tier: str | None = None
    sample_status: SampleStatus


# ---------------------------------------------------------------------------
# GET /api/v1/sample/{token}
# ---------------------------------------------------------------------------


class CollectionSample(BaseModel):
    """snake_case on the wire: ``doc_count``, ``sample_fields``."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    name: str
    doc_count: int
    sample_fields: list[str]


class SampleResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    collections: list[CollectionSample]


# ---------------------------------------------------------------------------
# POST /api/v1/plan
# ---------------------------------------------------------------------------


class ModuleName(str, Enum):
    FUNNEL = "funnel"
    COHORT = "cohort"
    RFM = "rfm"
    ATTRIBUTION = "attribution"
    ANOMALY = "anomaly"


class PlanRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str = Field(..., min_length=1)
    collections: list[str] = Field(..., min_length=1, max_length=20)
    modules: list[ModuleName] = Field(..., min_length=1, max_length=5)


class PlanStep(BaseModel):
    """One step in a plan. snake_case on the wire."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    module: ModuleName
    collection: str
    mql_pipeline: list[dict]
    estimated_runtime_s: float | None = None


class PlanResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    plan_id: str
    plan: list[PlanStep]


# ---------------------------------------------------------------------------
# POST /api/v1/render (SSE stream)
# ---------------------------------------------------------------------------


class RenderRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str = Field(..., min_length=1)
    plan_id: str = Field(..., min_length=1)


class RenderProgressData(BaseModel):
    model_config = ConfigDict(extra="forbid")

    completed: int
    total: int


class RenderDoneData(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    plan_id: str


# The SSE event types. The frontend expects an `event:` line that
# matches one of these four names. `data` is JSON-encoded.
RenderEventName = Literal["progress", "card", "error", "done"]


# ---------------------------------------------------------------------------
# GET /api/v1/dashboard
# ---------------------------------------------------------------------------


class LayoutItem(BaseModel):
    """One item in a react-grid-layout — index + position + size."""

    model_config = ConfigDict(extra="forbid")

    i: str
    x: int = Field(..., ge=0)
    y: int = Field(..., ge=0)
    w: int = Field(..., ge=1)
    h: int = Field(..., ge=1)


class DashboardLayout(BaseModel):
    """Per-breakpoint layouts for react-grid-layout."""

    model_config = ConfigDict(extra="forbid")

    lg: list[LayoutItem] = Field(default_factory=list)
    md: list[LayoutItem] = Field(default_factory=list)
    sm: list[LayoutItem] = Field(default_factory=list)
    xs: list[LayoutItem] = Field(default_factory=list)


class DashboardResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    layout: DashboardLayout
    cards: list[CardDescriptor]


# ---------------------------------------------------------------------------
# PUT /api/v1/dashboard/layout
# ---------------------------------------------------------------------------


class LayoutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str = Field(..., min_length=1)
    layout: DashboardLayout


class LayoutResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ok: bool


# ---------------------------------------------------------------------------
# POST /api/v1/cards
# ---------------------------------------------------------------------------


class AddCardRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str = Field(..., min_length=1)
    module: ModuleName
    collection: str | None = None
    params: dict | None = None


class AddCardResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    card_id: str
    card: CardDescriptor


# ---------------------------------------------------------------------------
# POST /api/v1/refresh
# ---------------------------------------------------------------------------


class RefreshRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    session_token: str = Field(..., min_length=1)


class RefreshResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    refreshed_count: int
    error_count: int


# ---------------------------------------------------------------------------
# GET /api/v1/health
# ---------------------------------------------------------------------------


class McpSubprocessHealth(BaseModel):
    """camelCase on the wire per Zod: ``mcpSubprocesses``."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    active: int
    max: int


class HealthStatus(str, Enum):
    OK = "ok"
    DEGRADED = "degraded"
    DOWN = "down"


class HealthResponse(BaseModel):
    """Mixed wire format: ``uptime_s`` is snake_case, ``mcp_subprocesses`` is the Python name but the alias is the wire name (Zod uses ``mcpSubprocesses``)."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    status: HealthStatus
    uptime_s: float
    version: str
    mcp_subprocesses: McpSubprocessHealth = Field(..., alias="mcpSubprocesses")


__all__ = [
    # Errors
    "ErrorCode",
    "ErrorBody",
    "ErrorResponse",
    # Connect
    "ConnectRequest",
    "ConnectResponse",
    "ConnectStatus",
    # Probe
    "ProbeResponse",
    "SampleStatus",
    # Sample
    "SampleResponse",
    "CollectionSample",
    # Plan
    "PlanRequest",
    "PlanResponse",
    "PlanStep",
    "ModuleName",
    # Render
    "RenderRequest",
    "RenderProgressData",
    "RenderDoneData",
    "RenderEventName",
    # Dashboard
    "DashboardResponse",
    "DashboardLayout",
    "LayoutItem",
    "LayoutRequest",
    "LayoutResponse",
    # Cards
    "AddCardRequest",
    "AddCardResponse",
    # Refresh
    "RefreshRequest",
    "RefreshResponse",
    # Health
    "HealthResponse",
    "HealthStatus",
    "McpSubprocessHealth",
]
